### PR TITLE
Add network restriction support

### PR DESCRIPTION
**Pull request by Don-Olu**
Resolves # . (This is applicable only if this pull request relates to a GitHub issue, delete the line otherwise)

Notes:
Introduces Keyvault Network access control list to prevent pubic access to the keyvaults.

### CODE REVIEW

**Comment by Don-Olu** [comment]
Test Build: 
https://sds-sandbox-build.platform.hmcts.net/job/HMCTS_Sandbox_PRE/job/pre-shared-infrastructure/job/demo/37/

File: README.md [hunk 1]
```diff
@@ -12,10 +12,20 @@ module "claim-store-vault" {
...
+  network_acls_allowed_subnet_id = [Jenkins Subnet id, other subnet id]
+  network_acls_allowed_ip_ranges = [Allowed ACL IPs]
+  network_acls_default_action = "Deny" or "Allow" # Deny by default
   common_tags         = var.common_tags
 }
 ```
 
+## Notes
+
+The module would create the follwoing permisions:
```
Comment (timja) [inline README.md:25]:
```suggestion
The module creates the following permissions:
```

File: README.md [hunk 2]
```diff
@@ -12,10 +12,20 @@ module "claim-store-vault" {
...
+ - Jenkins access to Keyvault
+ - Managed Identity ($product)-$env-mi
+ - Product team/developers access
```
Comment (timja) [inline README.md:28]:
```suggestion
 - Product team/developers access
 
```

File: main.tf [hunk 3]
```diff
@@ -17,6 +17,13 @@ resource "azurerm_key_vault" "kv" {
   enabled_for_template_deployment = true
   soft_delete_retention_days      = 90
 
+  network_acls {
+    bypass         = "AzureServices"
+    default_action = var.network_acls_default_action # Default is "Deny"
```
Comment (timja) [inline main.tf:22]:
```suggestion
    default_action = var.network_acls_default_action # Default is "Allow" for compatibility
```

File: variables.tf [hunk 4]
```diff
@@ -76,3 +76,19 @@ variable "soft_delete_enabled" {
...
 }
+
+variable "network_acls_allowed_ip_ranges" {
+  description = "IP Address space Allowed"
+  type        = list(string)
+  default     = []
+}
+
+variable "network_acls_default_action" {
+  default = "Deny"
```
Comment (timja) [inline variables.tf:87]:
```suggestion
  default = "Allow"
```

**Comment by Don-Olu** [comment]
private endpoints entails the use of peering and DNS modifications, it is reliant on other jobs outside Jenkins, 
using the above directly gives the terraform code the management of the ACL.

File: main.tf [hunk 5]
```diff
@@ -17,6 +17,13 @@ resource "azurerm_key_vault" "kv" {
   enabled_for_template_deployment = true
   soft_delete_retention_days      = 90
 
+  network_acls {
```
Comment (timja) [inline main.tf:20]: can you run `terraform fmt` on this file please? it looks unformatted

File: README.md [hunk 6]
```diff
@@ -12,10 +12,20 @@ module "claim-store-vault" {
   object_id           = var.jenkins_AAD_objectId
   resource_group_name = azurerm_resource_group.rg.name
   product_group_name  = "Your AAD group" # e.g. MI Data Platform, or dcd_cmc
+  create_managed_identity = true or false
+  network_acls_allowed_subnet_id = [Jenkins Subnet id, other subnet id]
+  network_acls_allowed_ip_ranges = [Allowed ACL IPs]
+  network_acls_default_action = "Deny" or "Allow" # Deny by default
```
Comment (timja) [inline README.md:18]: Please add to a separate example not the default one

**Review by timja — COMMENTED** [review]
I'll check this on Monday need to verify no impact to teams not using this feature

File: main.tf (hunk 3 above)
Comment (Don-Olu) [reply main.tf]: Default is changed to "Allow" for compatibility

File: README.md [hunk 7]
```diff
@@ -12,10 +12,20 @@ module "claim-store-vault" {
   object_id           = var.jenkins_AAD_objectId
   resource_group_name = azurerm_resource_group.rg.name
   product_group_name  = "Your AAD group" # e.g. MI Data Platform, or dcd_cmc
+  create_managed_identity = true or false
+  network_acls_allowed_subnet_id = [Jenkins Subnet id, other subnet id]
+  network_acls_allowed_ip_ranges = [Allowed ACL IPs]
+  network_acls_default_action = "Deny" or "Allow" # Allow by default
```
Comment (timja) [inline README.md:18]: needs to be moved to its own example like in https://github.com/hmcts/cnp-module-key-vault#application-access-using-managed-identities

File: README.md [hunk 8]
```diff
@@ -12,10 +12,20 @@ module "claim-store-vault" {
...
   common_tags         = var.common_tags
 }
 ```
 
+## Notes
+
+The module creates the follwoing permisions:
```
Comment (timja) [inline README.md:25]:
```suggestion
The module creates the following permissions:
```

File: main.tf (hunk 5 above)
Comment (Don-Olu) [reply main.tf:21]: terraform fmt has been executed

File: README.md (hunk 7 above)
Comment (timja) [reply README.md:18]: this is still pending

File: README.md [hunk 9]
```diff
@@ -62,6 +72,15 @@ Retrieve by name if you know the display name:
 $ az ad group list --query "[?displayName=='dcd_devops'].{DisplayName: displayName, ObjectID: objectId}" -o table
 ```
 
+## Keyvault access using Access Control List
+Allow the jenkins subnet id e.g [data.azurerm_subnet.jenkins_subnet.id] and others
+Allow the listed set of IPs
+```
+network_acls_allowed_subnet_id = [Jenkins Subnet id, other subnet id]
```
Comment (timja) [inline README.md:79]: this should include a method of getting the ID, https://github.com/hmcts/ccd-shared-infrastructure/blob/master/storage-account.tf#L65-L70

File: variables.tf [hunk 10]
```diff
@@ -73,6 +73,26 @@ variable "soft_delete_enabled" {
   description = "(deprecated) does nothing"
 }
 
+# variable "purge_protection" {
+#   default     = true
+#   description = "Purge Protection Enabled"
+# }
```
Comment (timja) [inline variables.tf:79]: why?

File: variables.tf [hunk 11]
```diff
@@ -73,6 +73,26 @@ variable "soft_delete_enabled" {
...
+  type        = list(string)
+  default     = []
+}
+
+variable "network_acls_default_action" {
+  default = "Allow"
+}
+
+variable "network_acls_allowed_subnet_id" {
+  description = "Subnet id allowed"
```
Comment (timja) [inline variables.tf:95]:
```suggestion
  description = "Allowed subnets"
```

File: variables.tf [hunk 12]
```diff
@@ -73,6 +73,26 @@ variable "soft_delete_enabled" {
...
+  description = "IP Address space Allowed"
+  type        = list(string)
+  default     = []
+}
+
+variable "network_acls_default_action" {
+  default = "Allow"
+}
+
+variable "network_acls_allowed_subnet_id" {
```
Comment (timja) [inline variables.tf:94]:
```suggestion
variable "network_acls_allowed_subnet_ids" {
```

**Review by timja — CHANGES_REQUESTED** [review]

**Review by timja — CHANGES_REQUESTED** [review]
nothing has changed between reviews

**Review by timja — APPROVED** [review]

**Pull request merged**